synapse/storage/databases/main/cache.py

```diff
@@ -253,6 +253,7 @@ def _invalidate_caches_for_event(
             self.get_applicable_edit.invalidate((relates_to,))
             self.get_thread_summary.invalidate((relates_to,))
             self.get_thread_participated.invalidate((relates_to,))
+            self.get_mutual_event_relations_for_rel_type.invalidate((relates_to,))
```

Comment:
Should the whole `_invalidate_caches_for_event` be moved to the base store class, alongside `_invalidate_state_caches`?

Another q: should this use `_attempt_to_invalidate_cache`?

Reply:
> Should the whole `_invalidate_caches_for_event` be moved to the base store class, alongside `_invalidate_state_caches`?

I'm curious, what benefits are you thinking of?

Reply:
Purely aesthetic - in my head at least there's a divide like:

- `databases/main/_base.py` - contains local cache invalidation helpers/wrappers
- `databases/main/cache.py` - contains the stream/database/distributed invalidation logic

Reply:
I think it makes sense to have `_invalidate_caches_for_event` and `_invalidate_state_caches` next to each other. And `_invalidate_caches_for_event` even uses `_attempt_to_invalidate_cache` now.

I don't have a strong opinion on which direction we should move things to though. Probably best to wait for a reviewer if they want to include that change here.

 - `synapse/storage/_base.py`
- `synapse/storage/databases/main/cache.py`

---

It looks like currently, that divide you mentioned lines up but that's totally determined by where `_attempt_to_invalidate_cache` and `_invalidate_state_caches` currently are in `synapse/storage/_base.py`.

By name, having all of the cache stuff be in `synapse/storage/databases/main/cache.py` seems like it would make sense 🤷

Reply:
> By name, having all of the cache stuff be in synapse/storage/databases/main/cache.py seems like it would make sense 🤷

Agreed! I did try this but ended up in a total mess of class resolution order so reverted 😬

Reply:
> total mess of class resolution order

Yeah, the CacheInvalidationWorkerStore is a real bottleneck for this, making it really hard to type-hint. (#11165, #11354,  #11271)